oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/async/ElasticResultRowAsyncIterator.java

```diff
@@ -111,6 +110,25 @@ public boolean hasNext() {
                 throw new IllegalStateException("Error reading next result from Elastic", e);
             }
         }
+
+        // Check if there are any exception traces filled from onFailure Callback in the errorQueue
+        // Any exception (such as ParseException) during the prefetch (init scanner) via the async call to ES would be available here
+        // when the cursor is actually being traversed.
+        // This is being done so that we can log the caller stack trace in case of any exception from ES and not just the trace of the async query thread.
+        if (!errorQueue.isEmpty()) {
+            Exception e = new Exception();
+            e.setStackTrace(Thread.currentThread().getStackTrace());
+            String exceptionMsg = null;
+            try {
+                exceptionMsg = errorQueue.take().getMessage();
```

Comment:
The call to `take()` has no timeout, so it can block for an unbounded time. This should not be a problem in this case, but as a general practice I think it's best to always have a timeout. Otherwise, there is a chance that a bug may cause the caller thread to block here forever, which can be very hard to debug as the system will seem to be hang without any useful information on the logs. So often, in these cases, the only way to figure out what is happening is to get a thread dump, which may not be easy to get. If all blocking operations have a timeout, then after a while the operation fails and we get an exception on the logs. 

In this case, as we checked that the queue has something in it, we could `poll()` instead, which fails immediately if the queue is empty.

Reply:
This should now be redundant since now I am using AtomicReference instead of queue